go/chat/replyto.go

```diff
@@ -0,0 +1,197 @@
+	if err != nil {
+		return nil, err
+	}
+	for _, msg := range msgs {
+		if msg != nil {
+			res = append(res, *msg)
+		}
+	}
+	return res, nil
+}
+
+func (f *ReplyFiller) validReplyTo(msgID *chat1.MessageID) bool {
+	return msgID != nil && *msgID > 0
+}
+
+func (f *ReplyFiller) getReplyTo(msg chat1.MessageUnboxed) *chat1.MessageID {
+	st, err := msg.State()
+	if err != nil {
+		return nil
+	}
+	switch st {
+	case chat1.MessageUnboxedState_VALID:
+		if msg.Valid().MessageBody.IsType(chat1.MessageType_TEXT) &&
+			f.validReplyTo(msg.Valid().MessageBody.Text().ReplyTo) {
+			return msg.Valid().MessageBody.Text().ReplyTo
+		}
+	case chat1.MessageUnboxedState_OUTBOX:
+		if msg.Outbox().PrepareOpts != nil && f.validReplyTo(msg.Outbox().PrepareOpts.ReplyTo) {
+			return msg.Outbox().PrepareOpts.ReplyTo
+		}
+	}
+	return nil
+}
+
+func (f *ReplyFiller) FillSingle(ctx context.Context, uid gregor1.UID, conv types.UnboxConversationInfo,
+	msg chat1.MessageUnboxed) (res chat1.MessageUnboxed, err error) {
+	msgs, err := f.Fill(ctx, uid, conv, []chat1.MessageUnboxed{msg})
+	if err != nil {
+		return res, err
+	}
+	if len(msgs) == 0 {
+		return msg, nil
+	}
+	return msgs[0], nil
+}
+
+func (f *ReplyFiller) Fill(ctx context.Context, uid gregor1.UID, conv types.UnboxConversationInfo,
+	msgs []chat1.MessageUnboxed) (res []chat1.MessageUnboxed, err error) {
+	defer f.Trace(ctx, func() error { return err }, "Fill")()
+	// Gather up the message IDs we need
+	repliedToMsgIDsLocal := newFillerReplyMsgs(f.localFetcher)
+	repliedToMsgIDsRemote := newFillerReplyMsgs(f.G().ConvSource.GetMessages)
+	for _, msg := range msgs {
+		st, err := msg.State()
+		if err != nil {
+			continue
+		}
+		switch st {
+		case chat1.MessageUnboxedState_VALID:
+			if msgID := f.getReplyTo(msg); msgID != nil {
+				if f.localOnly {
+					repliedToMsgIDsLocal.add(*msgID)
+				} else {
+					repliedToMsgIDsRemote.add(*msgID)
+				}
+			}
+		case chat1.MessageUnboxedState_OUTBOX:
+			if msgID := f.getReplyTo(msg); msgID != nil {
+				repliedToMsgIDsLocal.add(*msgID)
+			}
+		}
+	}
+
+	// Fetch messages
+	localMsgs, err := repliedToMsgIDsLocal.fill(ctx, uid, conv)
+	if err != nil {
+		f.Debug(ctx, "Fill: failed to get local messages: %s", err)
+		return res, err
+	}
+	remoteMsgs, err := repliedToMsgIDsRemote.fill(ctx, uid, conv)
+	if err != nil {
+		f.Debug(ctx, "Fill: failed to get remote messages: %s", err)
+		return res, err
+	}
+	origMsgs := append(localMsgs, remoteMsgs...)
+	transform := newBasicSupersedesTransform(f.G(), basicSupersedesTransformOpts{
+		UseDeletePlaceholders: true,
+	})
+	allMsgs, err := transform.Run(ctx, conv, uid, origMsgs)
+	if err != nil {
+		f.Debug(ctx, "Fill: failed to supersede replies: %s", err)
+		return res, err
+	}
+	replyMap := make(map[chat1.MessageID]chat1.MessageUnboxed)
+	for _, msg := range allMsgs {
+		replyMap[msg.GetMessageID()] = msg
+	}
+
+	// Modify messages
+	for _, msg := range msgs {
```

Comment:
why not replace a message at the index if there is a reply, cleans up the logic around making sure everything is appended and covers any future error cases so we never omit a message here

Reply:
Sure